### ISSUE TITLE
Research splitting offense and defense

### ISSUE THREAD

**Issue by sirgippy** [sirgippy/GippyRank4#40]
## Goal

Determine whether GippyRank’s current assumption of a single scalar latent team quality is materially limiting the model.

The core question is:

> **Does representing each team with separate offensive and defensive strength recover stable, useful information that the current one-dimensional team-quality model cannot express?**

This is an investigation into whether a future **Offense/Defense Posterior V2** is justified.

It is **not** a production implementation.

Do not change production rankings in this issue.

## Why

GippyRank currently represents each team-season with one latent quality distribution.

Conceptually:

```text
Team A quality ─┐
                ├── game margin
Team B quality ─┘
```

This means two teams with similar overall quality are treated as the same kind of latent object even if one is:

```text
elite offense
average defense
```

and the other is:

```text
average offense
elite defense
```

Recent investigations have repeatedly failed to produce stable gains by adding more detail to the individual game likelihood:

- conditional YPP
- primitive yards / plays
- richer margin uncertainty
- scoring-environment context
- blowout saturation

The within-season temporal investigation also found only a small signal that did not justify added latent-state complexity.

Before adding more football statistics, test a more fundamental structural hypothesis:

> **perhaps “team quality” is not one-dimensional.**

A separate offensive and defensive latent state may provide a more appropriate place for score and future box-score evidence to live.

## Core principle

Do not begin by adding EPA, success rate, yards, turnovers, or play-by-play.

First test whether the **scoreboard alone** provides evidence that offensive and defensive identities are distinct, persistent, and useful.

The initial observable information remains:

```text
points scored
points allowed
site
opponent
pairing
```

If an offense/defense decomposition cannot improve future prediction using the information already present in final scores, do not rescue it by immediately adding richer football data.

## Frozen production baseline

Preserve:

- Historical Likelihood V1
- Context Prior
- History Prior
- Posterior V1
- Performance V1
- weekly publication
- published rankings / PMFs
- current one-dimensional rank semantics

Production Posterior V1 is the comparison baseline.

The existing scalar team-quality model must remain available unchanged throughout the investigation.

## Temporal design

Use:

```text
training:     2008–2017
development:  2018–2021
evaluation:   2022–2025
```

An earlier training start may be used if existing data supports it cleanly without reconstructing unreliable historical inputs.

Use only 2018–2021 for candidate selection and architecture decisions.

Do not use 2022–2025 to choose model form or hyperparameters.

Do not use 2026 outcomes.

### Evaluation caveat

2022–2025 has already been examined extensively in previous GippyRank research.

Therefore:

> **2022–2025 is leakage-safe historical evaluation, but not untouched independent confirmation.**

Do not describe it as an independent test set.

---

# Stage 0: test whether offense and defense behave like separate persistent traits

Before fitting a two-dimensional posterior, establish whether offensive and defensive performance residuals have stable within-team structure.

## Residual definition

Construct leakage-safe game-level offensive and defensive residuals relative to the existing scalar V1 expectation.

For each team/game, conceptually define:

```text
offensive residual
    = points scored - expected points scored
```

and:

```text
defensive residual
    = expected opponent points - opponent points scored
```

so that:

```text
positive offensive residual = offense performed better than expected
positive defensive residual = defense performed better than expected
```

The exact expected-points decomposition must be defined explicitly and symmetrically.

Do not derive the residuals using future games.

If the current Historical Likelihood does not expose a clean expected-score decomposition, implement the simplest research-only decomposition consistent with:

- expected margin
- total scoring environment if required
- home / away / neutral site
- pairing

Do not silently assign the entire expected margin to either offense or defense.

Document the mathematical definition.

## Persistence tests

For each team-season, test at minimum:

```text
offense_t → offense_t+1
defense_t → defense_t+1
```

and compare with cross-component relationships:

```text
offense_t → defense_t+1
defense_t → offense_t+1
```

Report:

- lag-1 offensive residual correlation
- lag-2 offensive residual correlation
- lag-1 defensive residual correlation
- lag-2 defensive residual correlation
- offense-to-defense cross-lag correlations
- early-season vs late-season offensive residual relationship
- early-season vs late-season defensive residual relationship
- elapsed-time bins
- season-by-season results

Also report results after removing team-season mean residuals where appropriate so persistent scalar misestimation does not masquerade as component identity.

## Null controls

Use deterministic controls such as:

- shuffle game order within team-season
- shuffle offensive and defensive residual assignments across teams within season
- unrelated-team same-season pairings

The question is not merely whether a correlation is nonzero.

The question is:

> **Does a team’s offensive history specifically predict its own future offense, and its defensive history specifically predict its own future defense, beyond generic residual structure?**

## Stage 0 interpretation

Evidence supporting a meaningful offense/defense decomposition would include:

- offense persistence greater than shuffled/null controls
- defense persistence greater than shuffled/null controls
- same-component persistence stronger than cross-component persistence
- recurring effect across seasons
- meaningful effect sizes rather than only tiny statistically significant values

Evidence against it would include:

- near-zero same-component persistence
- offense and defense cross-signals as strong as same-component signals
- unstable signs
- behavior comparable to null controls

Do not build a full two-dimensional posterior solely because Stage 0 produces a technically significant result.

---

# Stage 1: two-component future-game diagnostic

If Stage 0 shows enough structure to justify continuing, implement a research-only model with separate offensive and defensive latent team states.

## Candidate structure

For each team `i`, define conceptually:

```text
O_i = offensive strength
D_i = defensive strength
```

For a game between teams `i` and `j`, model scoring using something structurally equivalent to:

```text
points_i ~ f(O_i, D_j, site, pairing)
points_j ~ f(O_j, D_i, site, pairing)
```

Exact distributional form is implementation judgment, subject to the constraints below.

## Required constraints

Keep the candidate deliberately simple.

Do not add:

- YPP
- yards
- plays
- turnovers
- EPA
- success rate
- quarterback data
- injuries
- drive data
- play-by-play

Do not introduce a large hyperparameter search.

Prefer a small frozen family such as:

```text
OD0 — independent offense / defense latent states
OD1 — same model with one explicitly predeclared shrinkage/correlation structure
```

Only add OD1 if there is a clear statistical reason to test correlation between offense and defense.

Do not invent additional variants after seeing development results.

## Priors

Context is primary.

The offense/defense prior must be derived from the existing scalar preseason prior without using target-season outcomes.

Prefer a deliberately neutral decomposition unless training data supports something stronger.

For example, the initial prior may center offense and defense around a common scalar preseason quality while allowing uncertainty in how that quality is split.

Do not create a hidden offseason offense/defense model from 2022–2025 outcomes.

History may be evaluated as a sensitivity check after development selection.

## Identifiability

Explicitly handle location/scale identifiability.

A model where:

```text
O_i + c
D_i + c
```

or another equivalent transformation leaves game probabilities unchanged is not acceptable.

Document and test the normalization / centering convention.

Examples may include:

- population-centered offensive and defensive effects
- fixed overall scoring intercept
- explicit sum-to-zero constraints

Exact method is implementation judgment.

---

# Stage 1 evaluation

Use leakage-safe historical cutoffs.

At every cutoff:

- infer only from games strictly before the cutoff
- predict games strictly after the cutoff
- use identical target keys across V1 and all OD candidates

Primary target:

```text
next game
```

Also report broader future-game results if inexpensive.

## Primary metrics

Because this investigation changes matchup structure rather than only rank ordering, prioritize future-game prediction.

Report:

```text
margin NLL
margin MAE
win Brier
```

If the candidate produces explicit team-score distributions, also report:

```text
home-score NLL / MAE
away-score NLL / MAE
total-points NLL / MAE
```

Use proper likelihood metrics only when the candidate defines the corresponding probability distribution honestly.

Do not manufacture pseudo-NLL metrics from point estimates.

## Breakdowns

Report:

- aggregate
- per season
- early / mid / late season
- favorite / underdog or expected-margin bins
- games involving strongly asymmetric inferred offense/defense teams where useful

Do not hand-pick teams for selection.

---

# Development gate

A two-component candidate earns historical evaluation only if it shows meaningful improvement over scalar Posterior V1 on 2018–2021.

At minimum require:

1. next-game margin NLL improvement of at least `0.010`
2. next-game margin MAE improvement of at least `0.10` points
3. win Brier worsening no greater than `0.001`
4. NLL improvement in at least `3 of 4` development seasons
5. no season NLL worsening greater than `0.015`

If explicit score likelihoods are available, require them not to show obvious calibration degradation.

Also require that inferred offense/defense states exhibit meaningful separation rather than simply duplicating the scalar latent variable.

For example, report:

- correlation between inferred offense and defense
- correlation of each component with scalar V1 quality
- variance of offense-minus-defense
- stability of asymmetric team profiles across cutoffs

## Stop condition

If no candidate clears the development gate:

> **stop**

Do not build a production offense/defense posterior.

Conclude that the current evidence does not justify splitting team quality into separate latent components.

---

# Candidate selection

If more than one predeclared candidate clears the gate:

1. choose the lowest aggregate development margin NLL
2. if candidates are within `0.002` NLL, prefer the simpler / more strongly regularized candidate
3. freeze all candidate settings before evaluating 2022–2025

Do not tune using evaluation results.

---

# Stage 2: historical evaluation

Only the single development-selected OD candidate proceeds to 2022–2025.

Compare against scalar Posterior V1 on identical scoring keys.

Use Context as primary.

History may be reported separately as sensitivity if reasonably inexpensive.

Report:

```text
future margin NLL
future margin MAE
win Brier
```

and, where supported:

```text
score NLL
score MAE
total-points calibration
```

Also report:

- per-season deltas
- early / mid / late results
- stability of offense/defense decomposition
- representative deterministic team profiles

---

# Final-rank evaluation

Future-game prediction is the first gate because offense/defense structure should primarily improve matchup modeling.

If the selected OD candidate survives Stage 2, then test whether it can support the existing GippyRank ranking product.

Do not assume a priori that:

```text
overall quality = offense + defense
```

or:

```text
overall quality = offense - defense
```

is the correct ranking quantity.

Investigate a small, predeclared mapping from the two-dimensional posterior to the existing overall quality/rank space.

The preferred interpretation is whichever is mathematically consistent with neutral-site expected performance against the population and can be evaluated without target leakage.

Compare final-rank outputs with Posterior V1 using:

```text
final-rank NLL
CRPS
expected-rank MAE
median-rank MAE
80% coverage
80% interval width
Top 5 / Top 10 / Top 25 Brier
```

Do not promote an offense/defense model solely because matchup prediction improves if the resulting ranking object is incoherent or materially worse calibrated.

---

# Recommendation classes

## A — offense/defense latent structure is warranted

Use when:

- Stage 0 shows distinct persistent offensive and defensive structure
- a frozen OD candidate clears the development gate
- 2022–2025 future-game evaluation remains directionally better
- improvement occurs in at least 3 of 4 evaluation seasons
- no catastrophic season appears
- offense/defense states are not merely redundant copies of scalar quality
- an overall ranking mapping can be defined cleanly without materially degrading rank calibration

This means:

> **open a separate production-design issue for Offense/Defense Posterior V2**

It does not mean:

> ship the research implementation directly

## B — real component structure, insufficient predictive gain

Use when offense/defense identities appear real but the two-component model does not improve future prediction enough to justify added complexity.

This result may still motivate revisiting the hypothesis later if richer offense/defense-specific evidence becomes available.

## C — retain scalar team quality

Use when:

- Stage 0 does not show convincing component persistence
- or the two-component candidate does not improve future-game prediction
- or the decomposition is unstable / poorly identified / redundant

Do not force A.

---

# Representative team examples

If Stage 1 runs, include a small deterministic set of examples such as:

- strongest inferred offense relative to overall quality
- strongest inferred defense relative to overall quality
- largest offense-minus-defense asymmetry
- largest V1-vs-OD prediction disagreement

For each show enough information to understand:

```text
scalar V1 quality
OD offensive estimate
OD defensive estimate
game prediction under V1
game prediction under OD
actual result
```

Examples are explanatory only.

They must not affect model selection.

---

# Potential confounders

Explicitly investigate:

## Pace / scoring environment

A high-scoring team may not necessarily have an unusually strong offense if its games simply contain more possessions.

Do not solve this by adding play count in v1.

Instead, report whether the scoreboard-only decomposition appears to absorb pace rather than actual offensive/defensive quality.

If this is a material limitation, document it as evidence for a future data investigation.

## Garbage time / blowouts

Do not introduce ad hoc garbage-time adjustments.

Historical Likelihood V1 remains the baseline interpretation of game evidence.

## Home / away

Site must remain explicit and symmetric.

Do not let home scoring inflate offense or road scoring suppress defense.

## Opponent quality

Every offensive observation is conditional on opponent defense and vice versa.

Do not calculate naïve points-per-game component ratings.

## FBS / FCS

Preserve existing FBS/FCS orientation and support semantics.

If an OD candidate cannot safely support FCS-FCS or FBS-FCS games, define an explicit fallback rather than silently dropping games.

---

# Computational guardrails

This investigation should be staged.

## Stage 0

Must be cheap.

## Stage 1

Use:

- small frozen candidate set
- development seasons only
- deterministic cutoff panel
- cached likelihood/scoring surfaces where practical

Do not brute-force dozens of offense/defense formulations.

## Stage 2

Only one frozen candidate proceeds.

If the model is too computationally expensive for ordinary historical evaluation, first investigate whether inference can be simplified or reused rather than launching an uncontrolled multi-hour candidate sweep.

---

# Tests

Add focused tests covering at least:

- offensive residual orientation
- defensive residual orientation
- home/away symmetry
- no future information in residual construction
- same-component vs cross-component diagnostic construction
- OD model identifiability constraint
- offensive game evidence depends on opponent defense
- defensive game evidence depends on opponent offense
- candidate grid is frozen
- only pre-cutoff games enter inference
- targets are strictly after cutoff
- comparison keys are identical across candidates
- development-only candidate selection
- 2022–2025 cannot influence candidate selection
- Context is primary
- FBS/FCS semantics remain valid
- scalar V1 behavior is unchanged
- production artifacts remain unchanged

Do not put the full historical experiment in the ordinary unit-test path.

---

# Artifacts

Write deterministic research artifacts under something like:

```text
data/processed/offense_defense_latent/
```

Include at least:

```text
report.md
summary.json
model_spec.json
stage0_component_persistence.csv
stage0_null_metrics.csv
development_future_metrics.csv
development_season_metrics.csv
development_component_diagnostics.csv
```

If Stage 2 runs:

```text
evaluation_future_metrics.csv
evaluation_season_metrics.csv
team_profiles.csv
```

If final-rank evaluation runs:

```text
rank_metrics.csv
rank_season_metrics.csv
```

Useful plots may include:

- offense persistence vs defense persistence
- same-component vs cross-component correlation
- inferred offense vs inferred defense
- V1 vs OD future-game NLL by season
- offense-minus-defense profile distribution

Do not generate decorative plots.

---

# Production integrity

This research must not change:

- Historical Likelihood V1
- Context / History priors
- production Posterior V1
- Performance V1
- weekly publication
- published rankings
- published PMFs
- publication slots
- site behavior

Hash-check relevant production artifacts before and after the research run.

---

# Validation

Run:

```bash
uv run pytest -q
uv run ruff check .
git diff --check
```

Run every triggered research stage to completion.

Repeat cheap deterministic stages where practical.

Do not rerun an extremely expensive final evaluation solely to claim byte-for-byte determinism if inputs and candidate selection are already frozen; document exactly what was replayed.

---

# Deliverables

Open a research PR and report:

- exact offense/defense residual definitions
- Stage 0 persistence and null-control results
- same-component vs cross-component effect sizes
- frozen OD candidate definitions
- identifiability convention
- preseason prior treatment
- 2018–2021 development metrics
- selected candidate or explicit `none`
- why selection did not use 2022–2025
- whether Stage 2 triggered
- 2022–2025 evaluation if triggered
- offense/defense profile stability
- representative deterministic team examples
- final-rank mapping/evaluation if reached
- final A/B/C recommendation
- runtime/artifact footprint
- production-integrity hashes
- tests/Ruff/diff-check
- PR URL/head/status

The central question is:

> **Is “team quality” really one thing, or does GippyRank need to understand that a team can be good in fundamentally different ways on offense and defense?**
